File: latexdraw-core/net.sf.latexdraw/src/main/net/sf/latexdraw/glib/models/impl/LGroupShape.scala
package net.sf.latexdraw.glib.models.impl

import java.awt.geom.Rectangle2D
import java.awt.Color
import scala.collection.JavaConversions.asScalaBuffer
import net.sf.latexdraw.glib.models.interfaces.shape.IGroup
import net.sf.latexdraw.glib.models.interfaces.shape.IPoint
import net.sf.latexdraw.glib.models.interfaces.shape.IShape
import net.sf.latexdraw.glib.models.ShapeFactory

/**
 * This trait encapsulates the code of the group related to the support of the general shape's properties.<br>
 * <br>
 * This file is part of LaTeXDraw.<br>
 * Copyright (c) 2005-2014 Arnaud BLOUIN<br>
 * <br>
 * LaTeXDraw is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * <br>
 * LaTeXDraw is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.<br>
 * <br>
 * 2012-04-16<br>
 * @author Arnaud BLOUIN
 * @since 3.0
 */
private[impl] trait LGroupShape extends IGroup {
	override def mirrorHorizontal(origin:IPoint) {
		getShapes.foreach{_.mirrorHorizontal(origin)}
	}

	override def mirrorVertical(origin:IPoint) {
		getShapes.foreach{_.mirrorVertical(origin)}
	}

	override def setThickness(thickness : Double) = getShapes.foreach{_.setThickness(thickness)}


	override def scale(sx : Double, sy : Double, pos : IShape.Position, bound : Rectangle2D) {
		getShapes.foreach{_.scale(sx, sy, pos, bound)}
	}


	override def getThickness: Double = {
		getShapes.find{_.isThicknessable} match {
			case Some(sh) => sh.getThickness
			case _ => Double.NaN
		}
	}


	override def isThicknessable = getShapes.exists{_.isThicknessable}


	override def isShowPtsable = getShapes.exists{_.isShowPtsable}


	override def isShowPts = getShapes.exists{shape => shape.isShowPtsable && shape.isShowPts}


	override def setShowPts(show : Boolean) {
		getShapes.filter{_.isShowPtsable}.foreach{_.setShowPts(show)}
	}


	override def getLineColour: Color = {
		getShapes.headOption match {
			case Some(shape) => shape.getLineColour
			case _ => Color.BLACK
		}
	}


	override def isLineStylable = getShapes.exists{_.isLineStylable}


	override def getLineStyle: IShape.LineStyle = {
		getShapes.find{_.isLineStylable} match {
			case Some(sh) => sh.getLineStyle
			case _ => IShape.LineStyle.SOLID
		}
	}


	override def setLineStyle(style : IShape.LineStyle) {
		getShapes.filter{_.isLineStylable}.foreach{_.setLineStyle(style)}
	}


	override def isBordersMovable = getShapes.exists{_.isBordersMovable}


	override def getBordersPosition: IShape.BorderPos = {
		getShapes.find{_.isBordersMovable} match {
			case Some(sh) => sh.getBordersPosition
			case _ => IShape.BorderPos.INTO
		}
	}


	override def setBordersPosition(position : IShape.BorderPos) {
		getShapes.filter{_.isBordersMovable}.foreach{_.setBordersPosition(position)}
	}


	override def setLineColour(lineColour : Color) {
		getShapes.foreach{_.setLineColour(lineColour)}
	}


	override def setDbleBordCol(colour : Color) {
		getShapes.filter{_.isDbleBorderable}.foreach{_.setDbleBordCol(colour)}
	}


	override def getDbleBordCol: Color = {
		getShapes.find{_.hasDbleBord} match {
			case Some(sh) => sh.getDbleBordCol
			case _ => Color.BLACK
		}
	}


	override def hasDbleBord = getShapes.exists{shape => shape.hasDbleBord && shape.isDbleBorderable }


	override def setHasDbleBord(dbleBorders : Boolean) {
		getShapes.filter{_.isDbleBorderable}.foreach{_.setHasDbleBord(dbleBorders)}
	}


	override def isDbleBorderable = getShapes.exists{_.isDbleBorderable}


	override def setDbleBordSep(dbleBorderSep : Double) {
		getShapes.filter{_.isDbleBorderable}.foreach{_.setDbleBordSep(dbleBorderSep)}
	}


	override def getDbleBordSep: Double = {
		getShapes.find{shape => shape.isDbleBorderable && shape.hasDbleBord} match {
			case Some(sh) => sh.getDbleBordSep
			case _ => Double.NaN
		}
	}


	override def isShadowable = getShapes.exists{_.isShadowable }


	override def hasShadow = getShapes.exists{shape => shape.isShadowable && shape.hasShadow }


	override def setHasShadow(shadow : Boolean) {
		getShapes.filter{_.isShadowable}.foreach{_.setHasShadow(shadow)}
	}


	override def setShadowSize(shadSize : Double) {
		getShapes.filter{_.isShadowable}.foreach{_.setShadowSize(shadSize)}
	}


	override def getShadowSize: Double = {
		getShapes.find{shape => shape.isShadowable && shape.hasShadow} match {
			case Some(sh) => sh.getShadowSize
			case _ => Double.NaN
		}
	}


	override def setShadowAngle(shadAngle : Double) {
		getShapes.filter{_.isShadowable}.foreach{_.setShadowAngle(shadAngle)}
	}


	override def getShadowAngle: Double = {
		getShapes.find{shape => shape.isShadowable && shape.hasShadow} match {
			case Some(sh) => sh.getShadowAngle
			case _ => Double.NaN
		}
	}


	override def setShadowCol(colour : Color) {
		getShapes.filter{_.isShadowable}.foreach{_.setShadowCol(colour)}
	}


	override def getShadowCol: Color = {
		getShapes.find{shape => shape.isShadowable && shape.hasShadow} match {
			case Some(sh) => sh.getShadowCol
			case _ => Color.BLACK
		}
	}


	override def getFillingStyle: IShape.FillingStyle = {
		getShapes.find{_.isInteriorStylable} match {
			case Some(sh) => sh.getFillingStyle
			case _ => IShape.FillingStyle.NONE
		}
	}


	override def setFillingStyle(style : IShape.FillingStyle) {
		getShapes.filter{_.isInteriorStylable}.foreach{_.setFillingStyle(style)}
	}


	override def isFillable = getShapes.exists{_.isFillable }


	override def isInteriorStylable = getShapes.exists{_.isInteriorStylable }


	override def setFillingCol(colour : Color) = {
		getShapes.filter{_.isFillable}.foreach{_.setFillingCol(colour)}
	}


	override def getFillingCol: Color = {
		getShapes.find{shape => shape.isFillable && shape.isFilled} match {
			case Some(sh) => sh.getFillingCol
			case _ => Color.BLACK
		}
	}


	override def setHatchingsCol(colour : Color) {
		getShapes.filter{_.isInteriorStylable}.foreach{_.setHatchingsCol(colour)}
	}


	override def getHatchingsCol: Color = {
		getShapes.find{shape => shape.isInteriorStylable && shape.getFillingStyle.isHatchings} match {
			case Some(sh) => sh.getHatchingsCol
			case _ => Color.BLACK
		}
	}


	override def setGradColStart(colour : Color) {
		getShapes.filter{_.isInteriorStylable}.foreach{_.setGradColStart(colour)}
	}


	override def getGradColStart: Color = {
		getShapes.find{shape => shape.isInteriorStylable && shape.getFillingStyle.isGradient} match {
			case Some(sh) => sh.getGradColStart
			case _ => Color.BLACK
		}
	}


	override def setGradColEnd(colour : Color) {
		getShapes.filter{_.isInteriorStylable}.foreach{_.setGradColEnd(colour)}
	}


	override def getGradColEnd: Color = {
		getShapes.find{shape => shape.isInteriorStylable && shape.getFillingStyle.isGradient} match {
			case Some(sh) => sh.getGradColEnd
			case _ => Color.BLACK
		}
	}


	override def setGradAngle(gradAngle : Double) {
		getShapes.filter{_.isInteriorStylable}.foreach{_.setGradAngle(gradAngle)}
	}


	override def getGradAngle: Double = {
		getShapes.find{shape => shape.isInteriorStylable && shape.getFillingStyle.isGradient} match {
			case Some(sh) => sh.getGradAngle
			case _ => Double.NaN
		}
	}


	override def setGradMidPt(gradMidPoint : Double) = {
		getShapes.filter{_.isInteriorStylable}.foreach{_.setGradMidPt(gradMidPoint)}
	}


	override def getGradMidPt: Double = {
		getShapes.find{shape => shape.isInteriorStylable && shape.getFillingStyle.isGradient} match {
			case Some(sh) => sh.getGradMidPt
			case _ => Double.NaN
		}
	}


	override def getHatchingsAngle: Double = {
		getShapes.find{shape => shape.isInteriorStylable} match {
			case Some(sh) => sh.getHatchingsAngle
			case _ => Double.NaN
		}
	}


	override def getHatchingsSep: Double = {
		getShapes.find{shape => shape.isInteriorStylable} match {
			case Some(sh) => sh.getHatchingsSep
			case _ => Double.NaN
		}
	}


	override def getHatchingsWidth: Double = {
		getShapes.find{shape => shape.isInteriorStylable} match {
			case Some(sh) => sh.getHatchingsWidth
			case _ => Double.NaN
		}
	}


	override def setHatchingsAngle(hatchingsAngle : Double) {
		getShapes.filter{_.isInteriorStylable}.foreach{_.setHatchingsAngle(hatchingsAngle)}
	}


	override def setHatchingsSep(hatchingsSep : Double) = {
		getShapes.filter{_.isInteriorStylable}.foreach{_.setHatchingsSep(hatchingsSep)}
	}


	override def setHatchingsWidth(hatchingsWidth : Double) = {
		getShapes.filter{_.isInteriorStylable}.foreach{_.setHatchingsWidth(hatchingsWidth)}
	}


	override def translate(tx : Double, ty : Double) = getShapes.foreach{_.translate(tx, ty)}


	override def addToRotationAngle(gravCentre : IPoint, angle : Double) = {
		val gc : IPoint = if(gravCentre==null) getGravityCentre else gravCentre
		getShapes.foreach{_.addToRotationAngle(gc, angle)}
	}


	override def setRotationAngle(rotationAngle : Double) = getShapes.foreach{_.setRotationAngle(rotationAngle)}


	override def rotate(point : IPoint, angle : Double) = getShapes.foreach{_.rotate(point, angle)}


	override def getRotationAngle: Double = {
		getShapes.headOption match {
			case Some(sh) => sh.getRotationAngle
			case _ => 0.0
		}
	}


	override def getGravityCentre: IPoint = {
		size match {
			case 0 => ShapeFactory.createPoint
			case _ => getTopLeftPoint.getMiddlePoint(getBottomRightPoint)
		}
	}


	override def getBottomRightPoint: IPoint = {
		var x : Double = Double.NaN
		var y : Double = Double.NaN

		if(size>0) {
			var br : IPoint = null
			x = Double.MinValue
			y = Double.MinValue

			getShapes.foreach{shape =>
				br = shape.getBottomRightPoint
				if(br.getX>x)
					x = br.getX
				if(br.getY>y)
					y = br.getY
			}
		}
		return ShapeFactory.createPoint(x, y)
	}


	override def getBottomLeftPoint: IPoint = {
		var x : Double = Double.NaN
		var y : Double = Double.NaN

		if(size>0) {
			var bl : IPoint = null
			x = Double.MaxValue
			y = Double.MinValue

			getShapes.foreach{shape =>
				bl = shape.getBottomLeftPoint
				if(bl.getX<x)
					x = bl.getX
				if(bl.getY>y)
					y = bl.getY
			}
		}
		return ShapeFactory.createPoint(x, y)
	}


	override def getTopLeftPoint: IPoint = {
		var x : Double = Double.NaN
		var y : Double = Double.NaN

		if(size>0) {
			var tl : IPoint = null
			x = Double.MaxValue
			y = Double.MaxValue

			getShapes.foreach{shape =>
				tl = shape.getTopLeftPoint
				if(tl.getX<x)
					x = tl.getX
				if(tl.getY<y)
					y = tl.getY
			}
		}
		return ShapeFactory.createPoint(x, y)
	}


	override def getTopRightPoint: IPoint = {
		var x : Double = Double.NaN
		var y : Double = Double.NaN

		if(size>0) {
			var tr : IPoint = null
			x = Double.MinValue
			y = Double.MaxValue

			getShapes.foreach{shape =>
				tr = shape.getTopRightPoint
				if(tr.getX>x)
					x = tr.getX
				if(tr.getY<y)
					y = tr.getY
			}
		}

		return ShapeFactory.createPoint(x, y)
	}


	override def hasHatchings = getShapes.exists{_.hasHatchings}


	override def hasGradient = getShapes.exists{_.hasGradient}
}

File: src/main/java/net/sf/latexdraw/models/impl/LGroupShape.java
/*
 * This file is part of LaTeXDraw.
 * Copyright (c) 2005-2018 Arnaud BLOUIN
 * LaTeXDraw is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * LaTeXDraw is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 */
package net.sf.latexdraw.models.impl;

import java.awt.geom.Rectangle2D;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import net.sf.latexdraw.models.ShapeFactory;
import net.sf.latexdraw.models.interfaces.shape.BorderPos;
import net.sf.latexdraw.models.interfaces.shape.Color;
import net.sf.latexdraw.models.interfaces.shape.FillingStyle;
import net.sf.latexdraw.models.interfaces.shape.IDot;
import net.sf.latexdraw.models.interfaces.shape.IGroup;
import net.sf.latexdraw.models.interfaces.shape.IPoint;
import net.sf.latexdraw.models.interfaces.shape.IShape;
import net.sf.latexdraw.models.interfaces.shape.ISquaredShape;
import net.sf.latexdraw.models.interfaces.shape.IStandardGrid;
import net.sf.latexdraw.models.interfaces.shape.LineStyle;
import net.sf.latexdraw.models.interfaces.shape.Position;
import net.sf.latexdraw.view.pst.PSTricksConstants;

/**
 * This trait encapsulates the code of the group related to the support of the general shape's properties.
 * @author Arnaud Blouin
 */
interface LGroupShape extends IGroup {
	@Override
	default void copy(final IShape sh) {
		//TODO
	}

	@Override
	default double getBorderGap() {
		return isEmpty() ? 0d : getShapes().get(0).getBorderGap();
	}

	@Override
	default double getDashSepBlack() {
		return getShapes().stream().filter(sh -> sh.isLineStylable()).findFirst().map(sh -> sh.getDashSepBlack()).orElse(Double.NaN);
	}

	@Override
	default double getDashSepWhite() {
		return getShapes().stream().filter(sh -> sh.isLineStylable()).findFirst().map(sh -> sh.getDashSepWhite()).orElse(Double.NaN);
	}

	@Override
	default double getDotSep() {
		return getShapes().stream().filter(sh -> sh.isLineStylable()).findFirst().map(sh -> sh.getDotSep()).orElse(Double.NaN);
	}

	@Override
	default IPoint getFullBottomRightPoint() {
		final double gap = getBorderGap();
		final IPoint br = getBottomRightPoint();
		br.translate(gap, gap);
		return br;
	}

	@Override
	default IPoint getFullTopLeftPoint() {
		final double gap = getBorderGap();
		final IPoint tl = getTopLeftPoint();
		tl.translate(-gap, -gap);
		return tl;
	}

	@Override
	default double getHeight() {
		return isEmpty() ? 0d : getShapes().get(0).getHeight();
	}

	@Override
	default int getNbPoints() {
		return 0;
	}

	@Override
	default ObservableList<IPoint> getPoints() {
		return FXCollections.emptyObservableList();
	}

	@Override
	default IPoint getPtAt(final int index) {
		return null;
	}

	@Override
	default double getWidth() {
		return isEmpty() ? 0d : getShapes().get(0).getWidth();
	}

	@Override
	default void scaleWithRatio(final double x, final double x2, final Position x3, final Rectangle2D x4) {
		//TODO ?
	}

	@Override
	default void setDashSepBlack(final double dash) {
		getShapes().stream().filter(sh -> sh.isLineStylable()).forEach(sh -> sh.setDashSepBlack(dash));
	}

	@Override
	default void setDashSepWhite(final double dash) {
		getShapes().stream().filter(sh -> sh.isLineStylable()).forEach(sh -> sh.setDashSepWhite(dash));
	}

	@Override
	default void setDotSep(final double dot) {
		getShapes().stream().filter(sh -> sh.isLineStylable()).forEach(sh -> sh.setDotSep(dot));
	}

	@Override
	default boolean shadowFillsShape() {
		return getShapes().stream().anyMatch(sh -> sh.shadowFillsShape());
	}

	@Override
	default void mirrorHorizontal(final double x) {
		getShapes().forEach(sh -> sh.mirrorHorizontal(x));
	}

	@Override
	default void mirrorVertical(final double y) {
		getShapes().forEach(sh -> sh.mirrorVertical(y));
	}

	@Override
	default void setThickness(final double thickness) {
		getShapes().forEach(sh -> sh.setThickness(thickness));
	}

	@Override
	default void scale(final double prevWidth, final double prevHeight, final Position pos, final Rectangle2D bound) {
		if(getShapes().stream().anyMatch(sh -> sh instanceof ISquaredShape || sh instanceof IStandardGrid || sh instanceof IDot)) {
			getShapes().forEach(sh -> sh.scaleWithRatio(prevWidth, prevHeight, pos, bound));
		}else {
			getShapes().forEach(sh -> sh.scale(prevWidth, prevHeight, pos, bound));
		}
	}

	@Override
	default double getThickness() {
		return getShapes().stream().filter(sh -> sh.isThicknessable()).map(sh -> sh.getThickness()).findFirst().orElse(Double.NaN);
	}

	@Override
	default double getFullThickness() {
		return getShapes().stream().filter(sh -> sh.isThicknessable()).map(sh -> sh.getFullThickness()).findFirst().orElse(Double.NaN);
	}

	@Override
	default boolean isColourable() {
		return getShapes().stream().anyMatch(sh -> sh.isColourable());
	}

	@Override
	default boolean isThicknessable() {
		return getShapes().stream().anyMatch(sh -> sh.isThicknessable());
	}

	@Override
	default boolean isShowPtsable() {
		return getShapes().stream().anyMatch(sh -> sh.isShowPtsable());
	}

	@Override
	default boolean isShowPts() {
		return getShapes().stream().anyMatch(sh -> sh.isShowPtsable() && sh.isShowPts());
	}

	@Override
	default void setShowPts(final boolean show) {
		getShapes().stream().filter(sh -> sh.isShowPtsable()).forEach(sh -> sh.setShowPts(show));
	}

	@Override
	default Color getLineColour() {
		return isEmpty() ? PSTricksConstants.DEFAULT_LINE_COLOR : getShapes().get(0).getLineColour();
	}

	@Override
	default boolean isLineStylable() {
		return getShapes().stream().anyMatch(sh -> sh.isLineStylable());
	}

	@Override
	default LineStyle getLineStyle() {
		return getShapes().stream().filter(sh -> sh.isLineStylable()).map(sh -> sh.getLineStyle()).findFirst().orElse(LineStyle.SOLID);
	}

	@Override
	default void setLineStyle(final LineStyle style) {
		getShapes().stream().filter(sh -> sh.isLineStylable()).forEach(sh -> sh.setLineStyle(style));
	}

	@Override
	default boolean isBordersMovable() {
		return getShapes().stream().anyMatch(sh -> sh.isBordersMovable());
	}

	@Override
	default BorderPos getBordersPosition() {
		return getShapes().stream().filter(sh -> sh.isBordersMovable()).map(sh -> sh.getBordersPosition()).findFirst().orElse(BorderPos.INTO);
	}

	@Override
	default void setBordersPosition(final BorderPos position) {
		getShapes().stream().filter(sh -> sh.isLineStylable()).forEach(sh -> sh.setBordersPosition(position));
	}

	@Override
	default void setLineColour(final Color lineColour) {
		getShapes().forEach(sh -> sh.setLineColour(lineColour));
	}

	@Override
	default void setDbleBordCol(final Color colour) {
		getShapes().stream().filter(sh -> sh.isDbleBorderable()).forEach(sh -> sh.setDbleBordCol(colour));
	}

	@Override
	default Color getDbleBordCol() {
		return getShapes().stream().filter(sh -> sh.hasDbleBord()).map(sh -> sh.getDbleBordCol()).findFirst().orElse(PSTricksConstants.DEFAULT_DOUBLE_COLOR);
	}

	@Override
	default boolean hasDbleBord() {
		return getShapes().stream().anyMatch(sh -> sh.isDbleBorderable() && sh.hasDbleBord());
	}

	@Override
	default void setHasDbleBord(final boolean dbleBorders) {
		getShapes().stream().filter(sh -> sh.isLineStylable()).forEach(sh -> sh.setHasDbleBord(dbleBorders));
	}

	@Override
	default boolean isDbleBorderable() {
		return getShapes().stream().anyMatch(sh -> sh.isDbleBorderable());
	}

	@Override
	default void setDbleBordSep(final double dbleBorderSep) {
		getShapes().stream().filter(sh -> sh.isLineStylable()).forEach(sh -> sh.setDbleBordSep(dbleBorderSep));
	}

	@Override
	default double getDbleBordSep() {
		return getShapes().stream().filter(sh -> sh.isDbleBorderable() && sh.hasDbleBord()).map(sh -> sh.getDbleBordSep()).findFirst().orElse(Double.NaN);
	}

	@Override
	default boolean isShadowable() {
		return getShapes().stream().anyMatch(sh -> sh.isShadowable());
	}

	@Override
	default boolean hasShadow() {
		return getShapes().stream().anyMatch(sh -> sh.isShadowable() && sh.hasShadow());
	}

	@Override
	default void setHasShadow(final boolean shadow) {
		getShapes().stream().filter(sh -> sh.isShadowable()).forEach(sh -> sh.setHasShadow(shadow));
	}

	@Override
	default void setShadowSize(final double shadSize) {
		getShapes().stream().filter(sh -> sh.isShadowable()).forEach(sh -> sh.setShadowSize(shadSize));
	}

	@Override
	default double getShadowSize() {
		return getShapes().stream().filter(sh -> sh.isShadowable() && sh.hasShadow()).map(sh -> sh.getShadowSize()).findFirst().orElse(Double.NaN);
	}

	@Override
	default void setShadowAngle(final double shadAngle) {
		getShapes().stream().filter(sh -> sh.isShadowable()).forEach(sh -> sh.setShadowAngle(shadAngle));
	}

	@Override
	default double getShadowAngle() {
		return getShapes().stream().filter(sh -> sh.isShadowable() && sh.hasShadow()).map(sh -> sh.getShadowAngle()).findFirst().orElse(Double.NaN);
	}

	@Override
	default void setShadowCol(final Color colour) {
		getShapes().stream().filter(sh -> sh.isShadowable()).forEach(sh -> sh.setShadowCol(colour));
	}

	@Override
	default Color getShadowCol() {
		return getShapes().stream().filter(sh -> sh.isShadowable() && sh.hasShadow()).
			map(sh -> sh.getShadowCol()).findFirst().orElse(PSTricksConstants.DEFAULT_SHADOW_COLOR);
	}

	@Override
	default FillingStyle getFillingStyle() {
		return getShapes().stream().filter(sh -> sh.isInteriorStylable()).map(sh -> sh.getFillingStyle()).findFirst().orElse(FillingStyle.NONE);
	}

	@Override
	default void setFillingStyle(final FillingStyle style) {
		getShapes().stream().filter(sh -> sh.isInteriorStylable()).forEach(sh -> sh.setFillingStyle(style));
	}

	@Override
	default void setFilled(final boolean filled) {
		getShapes().stream().filter(sh -> sh.isFillable()).forEach(sh -> sh.setFilled(filled));
	}

	@Override
	default boolean isFilled() {
		return getShapes().stream().anyMatch(sh -> sh.isFillable() && sh.isFilled());
	}

	@Override
	default boolean isFillable() {
		return getShapes().stream().anyMatch(sh -> sh.isFillable());
	}

	@Override
	default boolean isInteriorStylable() {
		return getShapes().stream().anyMatch(sh -> sh.isInteriorStylable());
	}

	@Override
	default void setFillingCol(final Color colour) {
		getShapes().stream().filter(sh -> sh.isFillable()).forEach(sh -> sh.setFillingCol(colour));
	}

	@Override
	default Color getFillingCol() {
		return getShapes().stream().filter(sh -> sh.isFillable() && sh.isFilled()).
			map(sh -> sh.getFillingCol()).findFirst().orElse(PSTricksConstants.DEFAULT_FILL_COLOR);
	}

	@Override
	default void setHatchingsCol(final Color colour) {
		getShapes().stream().filter(sh -> sh.isInteriorStylable()).forEach(sh -> sh.setHatchingsCol(colour));
	}

	@Override
	default Color getHatchingsCol() {
		return getShapes().stream().filter(sh -> sh.isInteriorStylable() && sh.getFillingStyle().isHatchings()).
			map(sh -> sh.getHatchingsCol()).findFirst().orElse(PSTricksConstants.DEFAULT_HATCHING_COLOR);
	}

	@Override
	default void setGradColStart(final Color colour) {
		getShapes().stream().filter(sh -> sh.isInteriorStylable()).forEach(sh -> sh.setGradColStart(colour));
	}

	@Override
	default Color getGradColStart() {
		return getShapes().stream().filter(sh -> sh.isInteriorStylable() && sh.getFillingStyle().isGradient()).
			map(sh -> sh.getGradColStart()).findFirst().orElse(PSTricksConstants.DEFAULT_GRADIENT_START_COLOR);
	}

	@Override
	default void setGradColEnd(final Color colour) {
		getShapes().stream().filter(sh -> sh.isInteriorStylable()).forEach(sh -> sh.setGradColEnd(colour));
	}

	@Override
	default Color getGradColEnd() {
		return getShapes().stream().filter(sh -> sh.isInteriorStylable() && sh.getFillingStyle().isGradient()).
			map(sh -> sh.getGradColEnd()).findFirst().orElse(PSTricksConstants.DEFAULT_GRADIENT_END_COLOR);
	}

	@Override
	default void setGradAngle(final double gradAngle) {
		getShapes().stream().filter(sh -> sh.isInteriorStylable()).forEach(sh -> sh.setGradAngle(gradAngle));
	}

	@Override
	default double getGradAngle() {
		return getShapes().stream().filter(sh -> sh.isInteriorStylable() && sh.getFillingStyle().isGradient()).
			map(sh -> sh.getGradAngle()).findFirst().orElse(Double.NaN);
	}

	@Override
	default void setGradMidPt(final double gradMidPoint) {
		getShapes().stream().filter(sh -> sh.isInteriorStylable()).forEach(sh -> sh.setGradMidPt(gradMidPoint));
	}

	@Override
	default double getGradMidPt() {
		return getShapes().stream().filter(sh -> sh.isInteriorStylable() && sh.getFillingStyle().isGradient()).
			map(sh -> sh.getGradMidPt()).findFirst().orElse(Double.NaN);
	}

	@Override
	default double getHatchingsAngle() {
		return getShapes().stream().filter(sh -> sh.isInteriorStylable()).map(sh -> sh.getHatchingsAngle()).findFirst().orElse(Double.NaN);
	}

	@Override
	default double getHatchingsSep() {
		return getShapes().stream().filter(sh -> sh.isInteriorStylable()).map(sh -> sh.getHatchingsSep()).findFirst().orElse(Double.NaN);
	}

	@Override
	default double getHatchingsWidth() {
		return getShapes().stream().filter(sh -> sh.isInteriorStylable()).map(sh -> sh.getHatchingsWidth()).findFirst().orElse(Double.NaN);
	}

	@Override
	default void setHatchingsAngle(final double hatchingsAngle) {
		getShapes().stream().filter(sh -> sh.isInteriorStylable()).forEach(sh -> sh.setHatchingsAngle(hatchingsAngle));
	}

	@Override
	default void setHatchingsSep(final double hatchingsSep) {
		getShapes().stream().filter(sh -> sh.isInteriorStylable()).forEach(sh -> sh.setHatchingsSep(hatchingsSep));
	}

	@Override
	default void setHatchingsWidth(final double hatchingsWidth) {
		getShapes().stream().filter(sh -> sh.isInteriorStylable()).forEach(sh -> sh.setHatchingsWidth(hatchingsWidth));
	}

	@Override
	default void translate(final double tx, final double ty) {
		getShapes().forEach(sh -> sh.translate(tx, ty));
	}

	@Override
	default void addToRotationAngle(final IPoint gravCentre, final double angle) {
		rotationAngleProperty().setValue(rotationAngleProperty().get() + angle);
		final IPoint gc = getGravityCentre();
		getShapes().forEach(sh -> sh.addToRotationAngle(gc, angle));
	}

	@Override
	default void setRotationAngle(final double rotationAngle) {
		rotationAngleProperty().setValue(rotationAngle);
		getShapes().forEach(sh -> sh.setRotationAngle(rotationAngle));
	}

	@Override
	default void rotate(final IPoint point, final double angle) {
		getShapes().forEach(sh -> sh.rotate(point, angle));
	}

	@Override
	default double getRotationAngle() {
		return size() == 1 ? getShapeAt(0).getRotationAngle() : rotationAngleProperty().doubleValue();
	}

	@Override
	default IPoint getGravityCentre() {
		return isEmpty() ? ShapeFactory.INST.createPoint() : getTopLeftPoint().getMiddlePoint(getBottomRightPoint());
	}

	@Override
	default IPoint getBottomRightPoint() {
		return getShapes().parallelStream().map(sh -> sh.getBottomRightPoint()).
			reduce((p1, p2) -> ShapeFactory.INST.createPoint(Math.max(p1.getX(), p2.getX()), Math.max(p1.getY(), p2.getY()))).
			orElseGet(() -> ShapeFactory.INST.createPoint(Double.NaN, Double.NaN));
	}

	@Override
	default IPoint getBottomLeftPoint() {
		return getShapes().parallelStream().map(sh -> sh.getBottomLeftPoint()).
			reduce((p1, p2) -> ShapeFactory.INST.createPoint(Math.min(p1.getX(), p2.getX()), Math.max(p1.getY(), p2.getY()))).
			orElseGet(() -> ShapeFactory.INST.createPoint(Double.NaN, Double.NaN));
	}

	@Override
	default IPoint getTopLeftPoint() {
		return getShapes().parallelStream().map(sh -> sh.getTopLeftPoint()).
			reduce((p1, p2) -> ShapeFactory.INST.createPoint(Math.min(p1.getX(), p2.getX()), Math.min(p1.getY(), p2.getY()))).
			orElseGet(() -> ShapeFactory.INST.createPoint(Double.NaN, Double.NaN));
	}

	@Override
	default IPoint getTopRightPoint() {
		return getShapes().parallelStream().map(sh -> sh.getTopRightPoint()).
			reduce((p1, p2) -> ShapeFactory.INST.createPoint(Math.max(p1.getX(), p2.getX()), Math.min(p1.getY(), p2.getY()))).
			orElseGet(() -> ShapeFactory.INST.createPoint(Double.NaN, Double.NaN));
	}

	@Override
	default boolean hasHatchings() {
		return getShapes().stream().anyMatch(sh -> sh.hasHatchings());
	}

	@Override
	default boolean hasGradient() {
		return getShapes().stream().anyMatch(sh -> sh.hasGradient());
	}
}
